chia/wallet/puzzles/cat.clvm

```diff
@@ -0,0 +1,407 @@
+;
+; This is all we need to do to ensure cats exactly balance in the inputs and outputs.
+;
+; Proof:
+;   Consider n, k, I_k values, O_k values, S_k and A_k as above.
+;   For the (CREATE_COIN_ANNOUNCEMENT Y_{k+1}) (created by the next coin)
+;   and (ASSERT_COIN_ANNOUNCEMENT hashed(Y_{k+1})) to match,
+;   we see that I_k can ensure that is has the correct value for S_{k+1}.
+;
+;   By induction, we see that S_{m+1} = sum(i, 1, m) [O_i - A_i] = sum(i, 1, m) O_i - sum(i, 1, m) A_i
+;   So S_{n+1} = sum(i, 1, n) O_i - sum(i, 1, n) A_i. But S_{n+1} is actually S_1 = 0,
+;   so thus sum(i, 1, n) O_i = sum (i, 1, n) A_i, ie. output total equals input total.
+
+;; GLOSSARY:
+;;  MOD_HASH: this code's sha256 tree hash
+;;  GENESIS_COIN_CHECKER: the function that determines if a coin can mint new cats
+;;  INNER_PUZZLE: an independent puzzle protecting the coins. Solutions to this puzzle are expected to
+;;              generate `AGG_SIG` conditions and possibly `CREATE_COIN` conditions.
+;; ---- items above are curried into the puzzle hash ----
+;;  inner_puzzle_solution: the solution to the inner puzzle
+;;  prev_coin_id: the id for the previous coin
+;;  this_coin_bundle: ((parent_id puzzle_hash amount) . lineage_proof)
+;;  next_coin_bundle: (parent_id inner_puzzle_hash amount)
+;;  prev_subtotal: the subtotal between prev-coin and this-coin
+;;
+;; coin_info: `(parent_id puzzle_hash amount)`. This defines the coin id used with ASSERT_MY_COIN_ID
+;; coin-bundle: the cons box `(coin_info . lineage_proof)`
+;;
+;; and automatically hashed in to the announcement generated with CREATE_COIN_ANNOUNCEMENT.
+;;
+
+(mod (MOD_HASH                 ;; curried into puzzle
+      LIMITATIONS_PROGRAM_HASH ;; curried into puzzle
+      INNER_PUZZLE             ;; curried into puzzle
+      inner_puzzle_solution    ;; if invalid, INNER_PUZZLE will fail
+      limitations_program      ;; The reveal of the hash curried above.  Can be () for default lineage checking.
+      prev_coin_id             ;; used in this coin's announcement, prev_coin ASSERT_COIN_ANNOUNCEMENT will fail if wrong
+      this_coin_bundle         ;; verified with ASSERT_MY_COIN_ID
+      next_coin_bundle         ;; used to generate ASSERT_COIN_ANNOUNCEMENT
+      prev_subtotal            ;; included in announcement, prev_coin ASSERT_COIN_ANNOUNCEMENT will fail if wrong
+      extra_delta              ;; this is the number of "legal discrepancy" between your real delta and what you're announcing your delta is
+    )
+
+     ;;;;; start library code
+
+     (include condition_codes.clvm)
+     (include curry-and-treehash.clinc)
+     (include cat_truths.clib)
+
+     (defmacro assert items
+         (if (r items)
+             (list if (f items) (c assert (r items)) (q . (x)))
+             (f items)
+         )
+     )
+
+     (defun-inline not (arg)
+      (i arg 0 1)
+     )
+
+     ; takes a lisp tree and returns the hash of it
+     (defun sha256tree1 (TREE)
+           (if (l TREE)
```

Comment:
Consider using the implementation in `sha256tree_module.clvm` (at the very least, rename this `sha256tree` since the operator version is now long-gone.